### PR TITLE
Reland "Improve the behavior of DropdownButton.disabledHint"

### LINKED ISSUE #71562: Improve the behavior of DropdownButton.disabledHint
`DropdownButton.disabledHint` was introduced by #18770

However, if a DropdownButton has some items and a selected value but the onChanged is null(disabled), the button only shows the [disabledHint] or [hint] or nothing if both of them are null. It should actually show the selected value instead of the hints, even if the button is disabled.

https://github.com/flutter/flutter/pull/70277 was the first attempt at this, but it failed some internal Google tests. At a glance, the tests seemed to be failing because a disabled dropdown button would display the selected value, but with the incorrect text color (it should be showing with the disabled color instead of the selected text color). https://github.com/flutter/flutter/pull/71559 reverted this PR.

cc/ @xu-baolin

### PR DESCRIPTION
**Pull request by xu-baolin**
## Description

This is relanding of #70277 

The diff with #70277:
Let `DefaultTextStyle` of selected item use `theme.disabledColor` when the button is diabled.

cc/: @shihaohong, Thanks very much for all your hard work on this change.

## Related Issues

Fixes #71562

## Tests

I added the following tests:

See files.

### CODE REVIEW

**Comment by xu-baolin** [comment]
@shihaohong I still have a question about the internal test,
before this change, the selected item does not render at all when the button is disabled, how did the test case check the selected item color?

File: packages/flutter/lib/src/material/dropdown.dart [hunk 1]
```diff
@@ -1351,7 +1357,7 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
     const Icon defaultIcon = Icon(Icons.arrow_drop_down);
 
     Widget result = DefaultTextStyle(
-      style: _textStyle!,
+      style: _enabled ? _textStyle! : _textStyle!.copyWith(color: Theme.of(context).disabledColor),
```
Comment (xu-baolin) [inline packages/flutter/lib/src/material/dropdown.dart:1360]: Here is the only change with the prior PR!

Comment (xu-baolin) [reply packages/flutter/lib/src/material/dropdown.dart:1360]: Of course I also add a test for this.

**Review by shihaohong — APPROVED** [review]
LGTM.

tldr; I think the apps worked around the problem that you have now fixed :) 
I think what was happening before was our code was working around the fact that when the hints are not defined with onChanged set to null, nothing shows up at all for the dropdown. However, the apps really wanted the currently selected value to be shown, but be disabled. So,  it was showing the selected value as the hint/disabledHint, which uses [`Theme.of(context).hintColor`](https://github.com/flutter/flutter/blob/94f7fef34f1c771f02204f047d2018d7010192bf/packages/flutter/lib/src/material/dropdown.dart#L1328). This PR should now fix that problem. 

This might still cause the diffs because hintColor and disabledColor are different by default, but this is a more acceptable change instead of using the enabled item colors.
